When running under Mac OS X or FreeBSD, temporarily block PROF si…

cinquin opened the pull request:
…gnal while a new thread is being spawned. This prevents the signal handler from being called before the new thread is fully initialized, and thus avoids crashes in GetEnv. The way this is implemented is by interposing calls to the native method Thread::start0. The PROF signal is unblocked in newly-created threads by the OnThreadStart callback.

I also found it necessary to only start profiling once the main thread has been created (as noted in the Java source for the Thread class, start and therefore start0 are not called for the main method thread).

This commit should not introduce any functional changes for platforms other than OS X and FreeBSD.

RichardWarburton commented:
I think overall this patch looks good to me. If anyone else has time to have a look over it 

Small nitpicks:

 * The indentation of functions is inconsistent with others in the file, would be good to be consistent.
 * Maybe its worth defining a compile time constant equivalent to `defined(__APPLE__) || defined(__FreeBSD__)` ? Seems like it would help to clarify the conditional. The real semantic condition we're compile-time branching on here is that the OS has a race between thread starting and safety of certain function calls. Seems like it should be a named concept.
 * Can we kill the goto please?

cinquin commented:
ping

RichardWarburton commented:
Hi @cinquin - the fixes to nitpicks look ok to me, only problem is that it has merge conflicts to fix. Can you resolve them and I'll merge. My apologies for the delay in replying.

cinquin commented:
I've rebased on your latest master. Please merge before before more conflicts arise!
I've added a commit to deal with crashes that occur when a thread in the process of stopping receives the PROF signal. But note for the record that I still see some infrequent crashes on thread exit, which I think are due to non-Java threads.

RichardWarburton commented:
Thanks.
